src/TwigEnvironmentFactory.php

```diff
@@ -74,9 +75,12 @@
 class TwigEnvironmentFactory
 {
     /**
-     * @throws Exception\InvalidConfigException for invalid config service values.
+     * @param ContainerInterface $container
+     *
+     * @return Environment
+     * @throws LoaderError
      */
-    public function __invoke(ContainerInterface $container) : TwigEnvironment
+    public function __invoke(ContainerInterface $container) : Environment
```

Comment:
For BC purposes, this should use the non-namespaced version in the signature. While the classes are equivalent, the namespaced version extends the non-namespaced one, so we need to keep the older one in the signature as it satisfies either return type.

It only needs to be reverted in this method, however; the other methods are private, so they do not affect the public API.

Reply:
actually, the [namespaced version](https://github.com/twigphp/Twig/blob/2.x/src/Environment.php) doesn't extend the [non-namespaced one](https://github.com/twigphp/Twig/blob/2.x/lib/Twig/Environment.php), its an alias . 
in the end of each non-namespaced class, [you can see the its an alias](https://github.com/twigphp/Twig/blob/2.x/lib/Twig/Environment.php#L970) and in the namespaced version class, it [imports the non-namespaced one](https://github.com/twigphp/Twig/blob/2.x/src/Environment.php#L5) that's why all classes are within an [unreachable statement ](https://github.com/twigphp/Twig/blob/2.x/src/Environment.php#L7) so this shouldn't cause any BC issues.

Reply:
I misread the first link (that's the file I'd found on which I'd based the argument that extension was occurring); the "extension" happens in a conditional that can never be triggered, while the alias happens instead in the class file defining `Twig_Environment` (I'm curious why they went this path, instead of doing the alias in the PSR-4 namespaced class file...).

That answers the BC questions for me, then. However, there are still the other issues to resolve.